File: backend/src/app/api/curation.py
"""Stage 1 — Curation: accept/reject decisions + canonical marking.

State lives in ``curation.json`` as ``{ recordKey → PhotoDecision }``. The
canonical rule (ROADMAP §6) is enforced here: exactly one canonical per specimen,
and only accepted photos can be canonical.
"""

from __future__ import annotations

from fastapi import APIRouter, HTTPException

from .. import models
from ..state import load_state, save_state
from . import deps

router = APIRouter(prefix="/api", tags=["curation"])


def _load(series: str) -> models.CurationState:
    raw = load_state(series, "curation")
    if raw is None:
        return models.CurationState(updated_at=deps.now_iso())
    return models.CurationState.model_validate(raw)


@router.get("/{series}/curation", response_model=models.CurationState)
def get_curation(series: str) -> models.CurationState:
    deps.get_series(series)
    return _load(series)


@router.get("/{series}/records", response_model=models.SeriesRecords)
def get_records(series: str) -> models.SeriesRecords:
    """Every parsed photo in the series (with ``specimenKey`` for grouping) plus the
    filenames the parser could not read. The curation grid groups ``records`` by
    ``specimenKey``; ``unparseable`` feeds the "couldn't parse these" panel.
    """
    ds = deps.get_series(series)
    return models.SeriesRecords(
        records=[deps.to_photo_model(r) for r in ds.records],
        unparseable=[u.filename for u in ds.unparseable],
    )


@router.put("/{series}/curation/{record_key:path}", response_model=models.CurationState)
def set_decision(
    series: str, record_key: str, decision: models.PhotoDecision
) -> models.CurationState:
    """Upsert one photo's decision, enforcing the canonical rule."""
    ds = deps.get_series(series)
    idx = deps.record_index(ds)
    if record_key not in idx:
        raise HTTPException(status_code=404, detail=f"Unknown record {record_key!r}.")

    if decision.is_canonical and decision.status != "accepted":
        raise HTTPException(
            status_code=400, detail="Only an accepted photo can be marked canonical."
        )

    st = _load(series)

    if decision.is_canonical:
        # Clear any prior canonical in the same specimen group (one per specimen).
        specimen_key = idx[record_key].specimen_key
        for rk, existing in st.photos.items():
            if rk == record_key:
                continue
            other = idx.get(rk)
            if other is not None and other.specimen_key == specimen_key and existing.is_canonical:
                existing.is_canonical = False

    st.photos[record_key] = decision
    st.updated_at = deps.now_iso()
    save_state(series, "curation", st)
    return st

File: frontend/src/api/curation.ts
// Stage 1 — Curation data hooks (ROADMAP Phase 3 step 1).
//
// Three hooks over the live backend:
//   • useSeriesRecords — GET /api/{series}/records  (photo grid data + unparseable)
//   • useCuration       — GET /api/{series}/curation (the decisions map)
//   • useSetDecision    — PUT /api/{series}/curation/{recordKey}, optimistic
//
// Request DTOs come from the generated `Schemas` (the OpenAPI contract), not
// domain.ts, so the wire shape can never drift from the server (CLAUDE.md).

import {
  useMutation,
  useQuery,
  useQueryClient,
  type UseQueryResult,
} from "@tanstack/react-query"
import { toast } from "sonner"

import { api, type Schemas } from "@/api/client"
import { queryKeys } from "@/api/hooks"
import { encodeRecordKey } from "@/api/util"
import type { CurationState, SeriesRecords } from "@/types/domain"

/** The PUT body — the decision to upsert. Typed from the API contract, not domain.ts. */
export type DecisionInput = Schemas["PhotoDecision"]

/* ─── Resource functions ─────────────────────────────────────────────────── */

function getSeriesRecords(seriesKey: string): Promise<SeriesRecords> {
  return api.get<SeriesRecords>(`/${encodeURIComponent(seriesKey)}/records`)
}

function getCuration(seriesKey: string): Promise<CurationState> {
  return api.get<CurationState>(`/${encodeURIComponent(seriesKey)}/curation`)
}

function putDecision(
  seriesKey: string,
  recordKey: string,
  decision: DecisionInput,
): Promise<CurationState> {
  return api.put<CurationState>(
    `/${encodeURIComponent(seriesKey)}/curation/${encodeRecordKey(recordKey)}`,
    decision,
  )
}

/* ─── Hooks ──────────────────────────────────────────────────────────────── */

/** Parsed source records for the grid. Disabled until a series is active. */
export function useSeriesRecords(
  seriesKey: string | null,
): UseQueryResult<SeriesRecords> {
  return useQuery({
    queryKey: queryKeys.records(seriesKey ?? "∅"),
    queryFn: () => getSeriesRecords(seriesKey as string),
    enabled: seriesKey != null,
  })
}

/** The curation decisions map (curation.json). Disabled until a series is active. */
export function useCuration(
  seriesKey: string | null,
): UseQueryResult<CurationState> {
  return useQuery({
    queryKey: queryKeys.curation(seriesKey ?? "∅"),
    queryFn: () => getCuration(seriesKey as string),
    enabled: seriesKey != null,
  })
}

interface SetDecisionVars {
  recordKey: string
  decision: DecisionInput
}

/**
 * Optimistic decision upsert. Clicking Accept/Reject/canonical reflects instantly;
 * on settle we invalidate both the curation cache and the stage-status cache so the
 * grid and the nav badges refresh (the 2B convention).
 *
 * The one-canonical-per-specimen rule is enforced authoritatively by the backend;
 * we mirror it optimistically (clearing sibling canonicals in the same specimenKey)
 * so the UI never briefly shows two canonicals before the server response lands.
 */
export function useSetDecision(seriesKey: string) {
  const qc = useQueryClient()
  const curationKey = queryKeys.curation(seriesKey)
  const recordsKey = queryKeys.records(seriesKey)

  return useMutation({
    mutationFn: ({ recordKey, decision }: SetDecisionVars) =>
      putDecision(seriesKey, recordKey, decision),

    onMutate: async ({ recordKey, decision }) => {
      await qc.cancelQueries({ queryKey: curationKey })
      const previous = qc.getQueryData<CurationState>(curationKey)

      // Which records share this photo's specimen? (for optimistic canonical clear)
      const records = qc.getQueryData<SeriesRecords>(recordsKey)?.records ?? []
      const specimenKey = records.find((r) => r.recordKey === recordKey)?.specimenKey
      const siblingKeys = new Set(
        records
          .filter((r) => r.specimenKey === specimenKey && r.recordKey !== recordKey)
          .map((r) => r.recordKey),
      )

      qc.setQueryData<CurationState>(curationKey, (old) => {
        const base: CurationState =
          old ?? { schemaVersion: 1, updatedAt: new Date().toISOString(), photos: {} }
        const photos = { ...base.photos }
        if (decision.isCanonical) {
          // Clear any prior canonical among this specimen's other photos.
          for (const rk of siblingKeys) {
            const sib = photos[rk]
            if (sib?.isCanonical) photos[rk] = { ...sib, isCanonical: false }
          }
        }
        photos[recordKey] = {
          status: decision.status,
          rejectionReason: decision.rejectionReason ?? null,
          isCanonical: decision.isCanonical,
          notes: decision.notes,
        }
        return { ...base, updatedAt: new Date().toISOString(), photos }
      })

      return { previous }
    },

    onError: (_err, _vars, ctx) => {
      // Roll back the optimistic write and surface the failure.
      if (ctx?.previous !== undefined) qc.setQueryData(curationKey, ctx.previous)
      toast.error("Couldn't save that change — please retry.")
    },

    onSettled: () => {
      void qc.invalidateQueries({ queryKey: curationKey })
      void qc.invalidateQueries({ queryKey: queryKeys.stageStatuses(seriesKey) })
    },
  })
}
